lib/Dialect/FIRRTL/Transforms/LowerGroups.cpp

```diff
@@ -0,0 +1,321 @@
+  return newModule;
+};
+
+/// Process amodule to remove any groups it has.
+void LowerGroupsPass::runOnModule(FModuleOp moduleOp) {
+  LLVM_DEBUG({
+    llvm::dbgs() << "Module: " << moduleOp.getModuleName() << "\n";
+    llvm::dbgs() << "  Examining Groups:\n";
+  });
+
+  CircuitOp circuitOp = moduleOp->getParentOfType<CircuitOp>();
+  StringRef circuitName = circuitOp.getName();
+
+  // Post-order traversal that expands a group into its parent. Foreach gorup
+  // found do the following:
+  //
+  // 1. Create and connect one ref-type output port for each value defined in
+  //    this group that drives an instance marked lowerToBind and move this
+  //    instance outside the group.
+  // 2. Create one input port for each value captured by this group.
+  // 3. Create a new module for this group and move the (mutated) body of this
+  //    group to the new module.
+  // 4. Instantiate the new module outside the group and hook it up.
+  // 5. Erase the group.
+  moduleOp.walk<mlir::WalkOrder::PostOrder>([&](Operation *op) {
+    auto group = dyn_cast<GroupOp>(op);
+    if (!group)
+      return WalkResult::advance();
+
+    // Compute the expanded group name.  For group @A::@B::@C, this is "A_B_C".
+    SmallString<32> groupName(group.getGroupName().getRootReference());
+    for (auto ref : group.getGroupName().getNestedReferences()) {
+      groupName.append("_");
+      groupName.append(ref.getValue());
+    }
+    LLVM_DEBUG(llvm::dbgs() << "    - Group: " << groupName << "\n");
+
+    Block *body = group.getBody(0);
+    SmallVector<PortInfo> ports;
+    SmallVector<Value> values;
+    OpBuilder builder(moduleOp);
+
+    // Handle instance ops marked lowerToBind.  These ops need to be moved
+    // outside the group to avoid nested binds which are illegal in the
+    // SystemVerilog specification (and checked by FIRRTL verification).
+    //
+    // For each value defined in this group which drives a port of one of these
+    // instances, create an output reference type port on the to-be-created
+    // module and drive it with the value.  Move the instance outside the group.
+    // We will hook it up later once we replace the group with an instance.
+    for (auto instanceOp :
+         llvm::make_early_inc_range(body->getOps<InstanceOp>())) {
+      // Instances which are not marked lowerToBind do not need to be split out.
+      if (!instanceOp.getLowerToBind())
+        continue;
+      LLVM_DEBUG({
+        llvm::dbgs()
+            << "      Found LowerToBind instance that needs to be moved:\n"
+            << "        module: " << instanceOp.getModuleName() << "\n"
+            << "        instance: " << instanceOp.getName() << "\n";
+      });
+      // Move the instance op outside the group.  Add an output port for any
+      // captures.
+      for (auto port : instanceOp.getResults()) {
+        auto connect = getSingleConnectUserOf(port);
+        if (!connect)
+          continue;
```

Comment:
Can we call signalPassFailure if ExpandWhen is assumed here?

Reply:
That's a good idea. 👍 This could also instead not use this utility and move all connections. 🤔  I'll see about making this work before `ExpandWhens`.